### PR TITLE
Throw an error on missing status in gRPC and gRPC-Web transports

### LINKED ISSUE #1108: RST_STREAM Handling
**Describe the bug**

When a grpc server is abruptly closed while a streaming RPC is open, client may see a RST_STREAM http/2 frame. No grpc level status, trailer, etc. are present in the response.

I'm using a grpc transport and a PromiseClient, which represents a stream as an async iterable. This scenario results in the client terminating the iterable without error, as though the stream was closed normally/successfully.

The [GRPC HTTP/2 protocol](https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md#errors) states the following:

> RPC runtime implementations should interpret RST_STREAM as immediate full-closure of the stream and should propagate an error up to the calling application layer.

In my particular case, which is using a fairly vanilla Istio/Envoy proxy, the HTTP/2 code returned is NO_ERROR(0). I see no grpc status header. 

I believe the expected client behaviour would be to throw an error from the iterable/stream so the application can handle it (possibly reconnecting, etc).

**To Reproduce**

I'm testing server restarts in a k8s deployment with Istio/Envoy in front of the services. Abruptly terminating pods, Envoy is responding with the RST_STREAM frame.

**Environment:**
- @connectrpc/connect version: `1.4.0`
- @connectrpc/connect-node version: `1.4.0`
- Node.js version: 20.12.2

### PR DESCRIPTION
**Pull request by srikrsna-buf**
Throw an error on missing status in gRPC and gRPC-Web transports. 

In node, there is no way to tell if a server terminated with `RST_STREAM` with code `NO_ERROR` or a regular close. This is causing such closures to not be reported. Checking to see if we received the trailers with the status code will result in the same case. This closes #1108.

This change also triggered a minor conformance failure of unsupported header check, so this PR also adds the check to gRPC transport. We continue to omit this for gRPC-Web for [compatibility](https://github.com/connectrpc/connect-es/pull/656) with other gRPC-Web libraries

### CODE REVIEW

File: packages/connect-node/src/transport.spec.ts [hunk 1]
```diff
@@ -0,0 +1,107 @@
+// Copyright 2021-2024 The Connect Authors
```
Comment (srikrsna-buf) [inline packages/connect-node/src/transport.spec.ts:1]: Couldn't find a good place to put this test, so created a new transport related test file.

File: packages/connect/src/protocol-grpc/validate-trailer.ts [hunk 2]
```diff
@@ -28,4 +31,12 @@ export function validateTrailer(trailer: Headers, header: Headers): void {
     });
     throw err;
   }
+  if (
+    !header.has(headerGrpcStatus) &&
+    !header.has(headerStatusDetailsBin) &&
+    !trailer.has(headerGrpcStatus) &&
+    !trailer.has(headerStatusDetailsBin)
+  ) {
+    throw new ConnectError("protocol error: missing status", Code.Internal);
```
Comment (timostamm) [inline packages/connect/src/protocol-grpc/validate-trailer.ts:40]: `validateResponse()` returns a `foundStatus: boolean` that only looks for the `grpc-status` field. I don't have a suggestion yet, but we should try to clean this up a bit.

File: packages/connect-node/src/transport.spec.ts [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    server: {
+      name: "Server",
+      I: Int32Value,
+      O: StringValue,
+      kind: MethodKind.ServerStreaming,
+    },
+  },
+} as const;
+
+describe("Calls should fail with code internal on RST_STREAM no_error before trailers are received", function () {
```
Comment (timostamm) [inline packages/connect-node/src/transport.spec.ts:37]: Could we possibly add this as a custom connectconformance test with the flag `--test-file`?

Comment (srikrsna-buf) [reply packages/connect-node/src/transport.spec.ts:37]: There is no way to trigger an `RST_STREAM(0x00)` using a test file. 

We can indirectly trigger it by providing a raw http response that ends the response before request is finished, but that also needs code changes in the conformance runner to allow that case. It [drains the requests](https://github.com/connectrpc/conformance/blob/368943e8bfff29a06fda135b8680433516dbcde5/internal/app/referenceserver/raw_response.go#L220-L228) before sending the raw response.

File: packages/connect/src/protocol-grpc/validate-trailer.ts (hunk 2 above)
Comment (timostamm) [reply packages/connect/src/protocol-grpc/validate-trailer.ts]: Discussed offline: We'll look into cleaning this up in v2.

**Review by timostamm — APPROVED** [review]

**Pull request merged**